src/test/java/com/hazelcast/commandline/member/MemberCommandLineTest.java

```diff
@@ -0,0 +1,193 @@
+package com.hazelcast.commandline.member;
+
+import com.hazelcast.commandline.CommandLineTestSupport;
+import com.hazelcast.commandline.HazelcastCommandLine;
+import com.hazelcast.core.LifecycleEvent;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import static junit.framework.TestCase.assertEquals;
+import static junit.framework.TestCase.assertTrue;
+
+public class MemberCommandLineTest extends CommandLineTestSupport {
```

Comment:
Writing tests for this tool is tricky, because they touch the operating system. There are a few problems with them:
1. They need an isolated environment (which you may not get and you don't specify explicitly that you need it)
2. They are neither acceptance tests, nor unit tests:
   * Acceptance would mean that you run from the user perspective, like
```
$ hazelcast member start
some-name
$ <check the logs some-name>
...
```
  * Unit tests would mean that you don't touch OS (or touch it in more controller way)

I don't have a clear solution how the test for this app should look like. Did you make any research how people test their CLI tools? 

Some ideas that could help;
1. Keep this file as an acceptance tests, but have here only tests that don't need to be isolated, like `test_start()` looks file. But `test_start_withJVMOptions() does not.
2. Think if you can change it to tests really from acceptance testing perspective. Like that the user could read it and understand.
3. Introduce some unit tests, it's hard, but you could have some abstraction over `Runtime.getRuntime().exec(command)`. Then, you could check the more specific cases not by check if you're running Hazelcast process is ok, but by checking the input for your abstraction. Think of it as you'd mock `Runtime.getRuntime().exec(command)`.

Reply:
I totally agree with your problem points. Those are where I spend my time in the beginning. 

1. Yes, we definitely need an isolated environment. This is what we discussed in the beginning.
2. I agree that these are not either acceptance tests, or unit tests. Here are the difficulties:

- Writing acceptance tests are very difficult when you spawn new processes from a JVM, because you need to place/access to the exact process that you spawned. It will always bring a third party accessor (such as REST) which will be another test point itself.
- Writing proper unit tests is another problem when you spawn new processes from a JVM, because you cannot mock the objects that you use in the application. Even you mock them, the testing criteria doesn't return from the mocks. It should be a mock of the running object, which makes the test non-sense.

> I don't have a clear solution how the test for this app should look like. Did you make any research how people test their CLI tools?

I have done some, but we can do more. 

Regarding your ideas;

1. I agree, we can remove `test_start_withJVMOptions` kind of tests. I had already not included tests for some features because of this reason.
2. I mentioned about the difficulties of acceptance tests above.
3. I mentioned about the difficulties of mocking above.

Please comment with your ideas, this is really helpful. 👍